testing/jupyterlab/markdown-reporter.ts

```diff
@@ -0,0 +1,130 @@
+// my-awesome-reporter.ts
+import fs from "fs";
+import path from "path";
+import type {
+  FullConfig,
+  Suite,
+  TestCase,
+  TestError,
+  Reporter,
+} from "@playwright/test/types/testReporter";
+
+class MyReporter implements Reporter {
+  config!: FullConfig;
+  rootSuite!: Suite;
+  private _errors: TestError[] = [];
+  private _outputFile: string | undefined;
+  private _testInfoPagesBaseURL: string;
+
+  constructor(
+    options: { outputFile?: string; testInfoPagesBaseURL: string }
+  ) {
+    this._outputFile = options.outputFile;
+    this._testInfoPagesBaseURL = options.testInfoPagesBaseURL;
+  }
+
+  printsToStdio() {
+    return !this._outputFile;
+  }
+
+  onBegin(config, suite) {
+    this.config = config;
+    this.rootSuite = suite;
+  }
+
+  // TODO: do something with the errors, format and output them to Markdown.
+  onError(error: TestError): void {
+    this._errors.push(error);
+  }
+
+  onEnd() {
+    outputReport(this._serializeReport(), this._outputFile);
+  }
+
+  private _serializeReport(): string {
+    const lines: string[] = [];
+    lines.push("# Test Results");
+    lines.push("");
+    const projectSuites = this.rootSuite.suites;
+
+    for (const projectSuite of projectSuites) {
+      const fileSuites = projectSuite.suites;
+      for (const fileSuite of fileSuites) {
+        lines.push(`## ${fileSuite.title}`);
+        lines.push("");
+        for (const test of fileSuite.allTests()) {
+          const testResultString = this._serializeTest(test);
+          lines.push(testResultString);
+        }
+        lines.push("");
+      }
+    }
+
+    return lines.join("\n");
+  }
+
+  private _serializeTest(test: TestCase): string {
+    const lines: string[] = [];
+    const outcome = {
+      expected: "✅",
+      unexpected: "❌",
+      skipped: "[skipped]",
+      flaky: "[flaky]",
+    }[test.outcome()];
+    const [_root, _project, file, ...titlePaths] = test.titlePath();
+    const title = titlePaths.join(" > ");
+    const infoURL = this.getTestInfoURL(test);
+    lines.push(`- ${outcome} [${title}](${infoURL})`);
```

Comment:
I'm not super happy right now with the way the Markdown report looks visually, but it can be improved later.

I also wonder if perhaps I should use a templating language like Handlebars to make it easier to edit the way the report looks. I mean, right now it's not so easy to get a sense of how the entire Markdown file will look just from looking at this file.

Reply:
Metadata.... such as?